akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -326,6 +315,23 @@ private[typed] class ActorCell[T](
         else unsafe.getAndAddInt(this, statusOffset, -1)
       }
     }
+
+    try {
+      unscheduleReceiveTimeout()
+      if (!isTerminated(status))
+        process()
+      scheduleReceiveTimeout()
+      afterProcess()
+    } catch {
+      case NonFatal(ex) ⇒
+        fail(ex)
+        afterProcess()
```

Comment:
Do you mean that the problem is that first `afterProcess` is called from the normal path and `afterProcess` itself throws an exception and then `afterProcess` is called again here?

Another ugly `var`, or do you see another way?

```scala
    var fatalError = false
    try {
      unscheduleReceiveTimeout()
      if (!isTerminated(status))
        process()
      scheduleReceiveTimeout()
    } catch {
      case NonFatal(ex) ⇒
        fail(ex)
      case ie: InterruptedException ⇒
        fail(ie)
        if (Debug) println(s"[$thread] $self interrupting due to catching InterruptedException")
        Thread.currentThread.interrupt()
      case t: Throwable =>
        fatalError = true
        throw t
    } finally {
      if (!fatalError) {
        // do the afterProcess stuff here
      }
    }
```